src/movement.h

```diff
@@ -132,29 +138,117 @@ class MoveEvent final : public Event
 		const std::string& getVocationString() const {
 			return vocationString;
 		}
+		void setVocationString(const std::string& str) {
+			vocationString = str;
+		}
 		uint32_t getWieldInfo() const {
 			return wieldInfo;
 		}
 		const VocEquipMap& getVocEquipMap() const {
 			return vocEquipMap;
 		}
+		void addVocEquipMap(std::string vocName) {
+			int32_t vocationId = g_vocations.getVocationId(vocName);
+			if (vocationId != -1) {
+				vocEquipMap[vocationId] = true;
+			}
+		}
+		bool getTileItem() const {
+			return tileItem;
+		}
+		void setTileItem(bool b) {
+			tileItem = b;
+		}
+		std::vector<uint32_t> getItemIdRange() {
+			return itemIdRange;
+		}
+		void addItemId(uint32_t id) {
+			itemIdRange.emplace_back(id);
+		}
+		std::vector<uint32_t> getActionIdRange() {
+			return actionIdRange;
+		}
+		void addActionId(uint32_t id) {
+			actionIdRange.emplace_back(id);
+		}
+		std::vector<uint32_t> getUniqueIdRange() {
+			return uniqueIdRange;
+		}
+		void addUniqueId(uint32_t id) {
+			uniqueIdRange.emplace_back(id);
+		}
+		std::vector<std::string> getPosList() {
+			return posList;
+		}
+		void addPosList(std::string pos) {
+			posList.emplace_back(pos);
+		}
+		std::string getSlotName() {
+			return slotName;
+		}
+		void setSlotName(std::string name) {
+			slotName = name;
+		}
+		void setSlot(uint32_t s) {
+			slot = s;
+		}
+		uint32_t getRequiredLevel() {
+			return reqLevel;
+		}
+		void setRequiredLevel(uint32_t level) {
+			reqLevel = level;
+		}
+		uint32_t getRequiredMagLevel() {
+			return reqMagLevel;
+		}
+		void setRequiredMagLevel(uint32_t level) {
+			reqMagLevel = level;
+		}
+		bool needPremium() {
+			return premium;
+		}
+		void setNeedPremium(bool b) {
+			premium = b;
+		}
+		uint32_t getWieldInfo() {
+			return wieldInfo;
+		}
+		void setWieldInfo(WieldInfo_t info) {
+			wieldInfo |= info;
+		}
 
-	private:
 		std::string getScriptEventName() const override;
 
+		static StepFunction StepInField;
```

Comment:
These statics seem to be unused?

Reply:
They're used in luascripts.cpp lines 13233-13248
```c++
        moveevent->stepFunction = moveevent->StepInField;
} else if (tmpStr == "stepout") {
	moveevent->setEventType(MOVE_EVENT_STEP_OUT);
	moveevent->stepFunction = moveevent->StepOutField;
} else if (tmpStr == "equip") {
	moveevent->setEventType(MOVE_EVENT_EQUIP);
	moveevent->equipFunction = moveevent->EquipItem;
} else if (tmpStr == "deequip") {
	moveevent->setEventType(MOVE_EVENT_DEEQUIP);
	moveevent->equipFunction = moveevent->DeEquipItem;
} else if (tmpStr == "additem") {
	moveevent->setEventType(MOVE_EVENT_ADD_ITEM);
	moveevent->moveFunction = moveevent->AddItemField;
} else if (tmpStr == "removeitem") {
	moveevent->setEventType(MOVE_EVENT_REMOVE_ITEM);
	moveevent->moveFunction = moveevent->RemoveItemField;
```

Reply:
Why are they static then?

Reply:
to be able to link them properly I had to make them static, if you have a better idea feel free to let me know

Reply:
I don't understand, you already have an instance of `MoveEvent` (`movevent`), so there's no need for them to be static, is there?

Also I don't think it even makes sense for them to be `static`, because it would limit any of these functions to just one per type?

Reply:
well for the userdata it's intended to just have an one type use
it would probably harm more if we could do something like this
```lua
local object = MoveEvent()
object.onEquip = function(...)
    return true
end
object.onAddItem = function(...)
   return false
end
```
This could crash the server easily with the wrong setup.

I'm completly open to whatever suggestion you have here.
However I've mostly just mimic'd the loadFunction from xml in this case, so the implementation is rather the same
```c++
bool MoveEvent::loadFunction(const pugi::xml_attribute& attr, bool isScripted)
{
	const char* functionName = attr.as_string();
	if (strcasecmp(functionName, "onstepinfield") == 0) {
		stepFunction = StepInField;
	} else if (strcasecmp(functionName, "onstepoutfield") == 0) {
		stepFunction = StepOutField;
	} else if (strcasecmp(functionName, "onaddfield") == 0) {
		moveFunction = AddItemField;
	} else if (strcasecmp(functionName, "onremovefield") == 0) {
		moveFunction = RemoveItemField;
	} else if (strcasecmp(functionName, "onequipitem") == 0) {
		equipFunction = EquipItem;
	} else if (strcasecmp(functionName, "ondeequipitem") == 0) {
		equipFunction = DeEquipItem;
	} else {
		if (!isScripted) {
			std::cout << "[Warning - MoveEvent::loadFunction] Function \"" << functionName << "\" does not exist." << std::endl;
			return false;
		}
	}

	if (!isScripted) {
		scripted = false;
	}
	return true;
}
```
The only difference was that with the functions having a namespace I couldn't call them from another cpp file which gave me trouble

Reply:
Ehh, I just realized these are member function not variables... Can you change the declaration so that it's less confusing and actually looks like a function declaration?
```suggestion
		static uint32_t StepInField(Creature* creature, Item* item, const Position& pos);
```

Reply:
couldn't we get rid of all this then?
```c++
typedef uint32_t(StepFunction)(Creature* creature, Item* item, const Position& pos);
typedef uint32_t(MoveFunction)(Item* item, Item* tileItem, const Position& pos);
typedef uint32_t(EquipFunction)(MoveEvent* moveEvent, Player* player, Item* item, slots_t slot, bool boolean);
```

Reply:
It's still used to define the callback pointer data members:
```		StepFunction* stepFunction;
		MoveFunction* moveFunction;
		EquipFunction* equipFunction;```
If they're used elsewhere still, they should stay, if not, I don't mind if they're removed.

Reply:
fixed in [c73ed97](https://github.com/otland/forgottenserver/pull/2558/commits/c73ed9783e0c33312326a790475cd4eee4f61ab9)